kernel_tuner/observers/ncu.py

```diff
@@ -0,0 +1,36 @@
+from kernel_tuner.observers import PrologueObserver
+
+try:
+    import nvmetrics
+except (ImportError):
+    nvmetrics = None
+    pass
+
+class NCUObserver(PrologueObserver):
+    """``NCUObserver`` measures performance counters.
+
+    """
+
+    def __init__(self, metrics=None):
+        """Create a new ``NCUObserver``.
+
+        :param metrics: The metrics to observe. This should be a list of strings.
+                        You can use ``ncu --query-metrics`` to get a list of valid metrics.
```

Comment:
Great! Here is an idea: Would it be much work to add a way to get the list of supported metrics from `nvmetrics`? Something like `NCUObserver.get_available_metrics()` could be useful to check which metrics are supported by the current GPU. Not a critical issue, just a thought

Reply:
A simple way to implement this is to let `NCUObserver` call the `ncu` command-line profiler with `--query-metrics`. It would arguably be better if `nvmetrics` could do this, as it links to CUPTI etc. anyway. I will try to figure out whether this is possible.

I have also thought about making this observer more user-friendly by providing presets of metrics. It is very tricky, though, as for many metrics there are mandatory suffixes like `.sum` or `.average`. We don't know what kind of metrics users are interested in. Moreover, the available metrics may differ per architecture. I think that this observer should mainly be used by expert users that profiled their code before (using `ncu` or the GUI) and found some metrics that they would like to know for a larger parameter space.

Reply:
Aren't there at least some metrics that are available on all GPUs that could be listed? I mean to get people started using the NCUObserver at first, and then if they want more they can still query ncu --query-metrics.

Reply:
I have added the functionality to query metrics to `nvmetrics`, including a Python interface. You can now do:
```
nvmetrics.queryMetrics(nvmetrics.NVPW_METRIC_TYPE_COUNTER)
nvmetrics.queryMetrics(nvmetrics.NVPW_METRIC_TYPE_RATIO)
nvmetrics.queryMetrics(nvmetrics.NVPW_METRIC_TYPE_THROUGHPUT)
```

These different metric types should be passed with some suffix, these are called `NVPW_RollupOp` and `NVPW_Submetric`, but not all combinations are valid, see this struct:
```
    typedef struct NVPW_MetricEvalRequest
    {
        /// the metric index as in 'NVPW_MetricsEvaluator_GetMetricNames'
        size_t metricIndex;
        /// one of 'NVPW_MetricType'
        uint8_t metricType;
        /// one of 'NVPW_RollupOp', required for Counter and Throughput, doesn't apply to Ratio
        uint8_t rollupOp;
        /// one of 'NVPW_Submetric', required for Ratio and Throughput, optional for Counter
        uint16_t submetric;
    } NVPW_MetricEvalRequest;
```

Reply:
@stijnh, how would you suggest proceeding here?
We could for instance use the same metrics that the example uses as default.